### DISCUSSION TITLE
Annotated generic

### DISCUSSION THREAD

**Discussion by 9en9i** [sqlalchemy/sqlalchemy#11189] [Usage Questions]
Hi, I often use a construct to specify primary keys in tables:

```python
T = TypeVar("T")

Primary: TypeAlias = Annotated[T, mapped_column(primary_key=True)]
```

But found that it doesn't work if the code looks like this:

```python
from typing import Annotated, TypeAlias, TypeVar
from sqlalchemy import BIGINT
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


T = TypeVar("T")
BigInt = Annotated[int, mapped_column(BIGINT())]

Primary: TypeAlias = Annotated[T, mapped_column(primary_key=True)]
Unique: TypeAlias = Annotated[T, mapped_column(unique=True)]

class BaseModel(DeclarativeBase):
    pass

class SomeModel(BaseModel):
    __tablename__ = 'some_table'

    id: Mapped[Primary[BigInt]]
    some_field: Mapped[Unique[BigInt]]
```

Here I get the error `sqlalchemy.exc.ArgumentError: Mapper Mapper[SomeModel(some_table)] could not assemble any primary key columns for mapped table 'some_table'`. 
Besides, if I still specify `id: Mapped[Primary[int]]`, I get this code (without using `UniqueConstraint`) during autogeneration of alembic migration:

```python
op.create_table(
    "some_table",
    sa.Column("id", sa.Integer(), nullable=False),
    sa.Column("some_field", sa.BIGINT(), nullable=False),
    sa.PrimaryKeyConstraint("id"),
)
```

Additionally, I saw that there are now changes in the main branch to use a new construct in python 3.12 - `type`. I installed sqlalchemy from main to test such code:

```python
from typing import Annotated
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


type Primary[T] = Annotated[T, mapped_column(primary_key=True)]

class BaseModel(DeclarativeBase):
    pass

class SomeModel(BaseModel):
    __tablename__ = 'some_table'

    id: Mapped[Primary[int]]
```

But I get the following error: `sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type Primary[int] inside the 'id' attribute Mapped annotation`.

Are there any plans to fix this behavior?

**Comment by zzzeek** [comment]
hi -

Please dont use the word "fix" since you are presenting an entirely new and novel use case that's never been proposed; that of **nesting** `Annotated` types within each other, and having them cascade onto each other.   SQLAlchemy is not a static type checker, it interprets pep-484 annotations at runtime in specific, documented ways, so this is not a "bug".   it's an idea for a new use case.

The type you provide in the first example presents this construct at runtime to SQLAlchemy:

```
(Pdb) our_type
typing.Annotated[int, <sqlalchemy.orm.properties.MappedColumn object at 0x7fc1497d9d80>, <sqlalchemy.orm.properties.MappedColumn object at 0x7fc1497d9e10>]

```

two MappedColumns.  So it is theoretically possible that the resolution logic could be enhanced to cascade these two elements onto one another.

in the second case we get something totally different:

```
(Pdb) our_type
Primary[int]

```

I have to use a complex unwrapping scheme to get to the mapped column here.   there is risk in these unwrapping schemes since they can conflict with others, this one seems to get to it

```
(Pdb) our_original_type
Primary[int]
(Pdb) get_args(our_original_type)
(<class 'int'>,)
(Pdb) our_original_type.__origin__
Primary
(Pdb) our_original_type.__origin__.__value__
typing.Annotated[T, <sqlalchemy.orm.properties.MappedColumn object at 0x7f24c26bf880>]
(Pdb) get_args(our_original_type.__origin__.__value__)
(T, <sqlalchemy.orm.properties.MappedColumn object at 0x7f24c26bf880>)

```

I would say these are both very complicated, novel uses that could be added but they would take some effort.

**Reply by 9en9i** [reply]
Thank you for your reply. I didn't open the issue to clarify maybe it wasn't intended. It just really seems like the documentation doesn't define the scope of using `Annotated`, so after a long time of using the `Primary[int]` construct I was surprised that `Primary[BigInt]` doesn't work. And my example seems to me to look intuitively natural and very concise. It would be cool if it worked like that.

**Reply by zzzeek** [reply]
we can work with PRs for things like this.    These are two new behaviors that we could add to the test suite and within the area that runtime-interprets types.

**Reply by 9en9i** [reply]
Yeah, okay, I'll try to open a PR soon

**Reply by 9en9i** [reply]
@zzzeek 
Hi, I've written some code, but I'm not ready to open a PR yet. I want to add parameterization to the test and also clarify one point. When type or other argument of mapped_column is defined more than once, should an exception be thrown or just a warning?

**Reply by zzzeek** [reply]
> When type or other argument of mapped_column is defined more than once

defined where exactly?   I dont understand what that means

**Reply by 9en9i** [reply]
I'm talking about a case like this:
```python
T = TypeVar("T")
BigInt = Annotated[int, mapped_column(BIGINT())]
SomeName = Annotated[T, mapped_column(unique=True)]
SomeName2 = Annotated[T, mapped_column(unique=False)]


class SomeModel(BaseModel):
    __tablename__ = "some_table"

    id: Mapped[BigInt] = mapped_column(Integer(), primary_key=True)
    field: Mapped[SomeName2[SomeName[int]]]
```

**Reply by zzzeek** [reply]
I think there should be a resolution order, if you get a series of mapped_column() constructs unpacked from your type, then you'd have to merge each one in series into your target column, going from the most general to the most specific, based on how the type is constructed.   emitting a warning for an ambiguity like the one above is probably more difficult to implement since you'd have to somehow catch for all such parameters that conflict.  I think above the SomeName2 is more specific to what is wanted in the mapping, so SomeName applies first, then SomeName2.

**Reply by 9en9i** [reply]
Now I have made the mapped_column() series to be mapped into a new MappedColumn selecting elements in order from types and if the argument has already been set an ArgumentError exception is thrown. and I think whether to leave it (for less ambiguity) or to replace it with warning

**Reply by CaselIT** [reply]
>  I think above the SomeName2 is more specific to what is wanted in the mapping, so SomeName applies first, then SomeName2.

IIRC python collapses Annotated into a single one, so I would follow the same resolution order used by python:
```py
import typing as T
a = T.Annotated[int, 'a']
b = T.Annotated[a, 'b']
c = T.Annotated[b, 'c']
print(c) # typing.Annotated[int, 'a', 'b', 'c']
print(T.get_args(c)) # (int, 'a', 'b', 'c')
```
